pumpkin/src/command/commands/cmd_msg.rs

```diff
@@ -0,0 +1,69 @@
+use async_trait::async_trait;
+use pumpkin_core::text::{color::NamedColor, TextComponent};
+
+use crate::command::{
+    args::{arg_message::MsgArgConsumer, arg_players::PlayersArgumentConsumer, Arg, ConsumedArgs},
+    tree::CommandTree,
+    tree_builder::argument,
+    CommandExecutor, CommandSender, InvalidTreeError,
+};
+use InvalidTreeError::InvalidConsumptionError;
+
+const NAMES: [&str; 3] = ["msg", "tell", "w"];
+
+const DESCRIPTION: &str = "Sends a private message to one or more players.";
+
+const ARG_TARGET: &str = "target";
+const ARG_MESSAGE: &str = "message";
+
+struct MsgExecutor;
+
+#[async_trait]
+impl CommandExecutor for MsgExecutor {
+    async fn execute<'a>(
+        &self,
+        sender: &mut CommandSender<'a>,
+        _server: &crate::server::Server,
+        args: &ConsumedArgs<'a>,
+    ) -> Result<(), InvalidTreeError> {
+        let sender_name = match sender {
+            CommandSender::Console => "Console",
+            CommandSender::Rcon(_) => "Rcon",
+            CommandSender::Player(player) => &player.gameprofile.name.clone(),
+        };
+
+        let Some(Arg::Players(targets)) = args.get(&ARG_TARGET) else {
+            return Err(InvalidConsumptionError(Some(ARG_TARGET.into())));
+        };
+
+        let Some(Arg::Msg(msg)) = args.get(ARG_MESSAGE) else {
+            return Err(InvalidConsumptionError(Some(ARG_MESSAGE.into())));
+        };
+
+        for target_player in targets {
+            let recipient_message = format!("{sender_name} whispers to you: {msg}");
+            let sender_message =
+                format!("you whisper to {}: {msg}", target_player.gameprofile.name);
+
+            let recipient_text = TextComponent::text(&recipient_message)
+                .color_named(NamedColor::Gray)
+                .italic();
+
+            let sender_text = TextComponent::text(&sender_message)
+                .color_named(NamedColor::Gray)
+                .italic();
```

Comment:
You could lift string formatting out of the loop so it isn't done again for every single target

The sender should receive only one "you whisper to [...]" message even if there are multiple targets, so you should move that out of the loop too

Reply:
Thanks, I hadn't even thought about that, everything changed ✅, I don't know if I did it in the best way, I'm still new to Rust 💪